File: solutions/2026-08-31/javascript/2026-08-31-find-the-minimum-and-maximum-number-of-nodes-between-critical-points.js
/*
 * Problem: Find the Minimum and Maximum Number of Nodes Between Critical Points
 * Link: https://leetcode.com/problems/find-the-minimum-and-maximum-number-of-nodes-between-critical-points/
 *
 * Approach Explanation:
 * The problem requires finding local maxima and minima (critical points) in a linked list
 * and then calculating the minimum and maximum distances between any two distinct critical points.
 *
 * 1. Initialization:
 *    - Initialize `prevNode` to `head`, `currNode` to `head.next`, and `index` to 1 (for 0-based indexing of nodes).
 *    - `firstCriticalPointIndex`: Stores the index of the very first critical point found. Initialized to -1.
 *    - `prevCriticalPointIndex`: Stores the index of the most recently found critical point. Initialized to -1.
 *    - `minDistance`: Stores the minimum distance between adjacent critical points. Initialized to `Infinity`.
 *    - `maxDistance`: Stores the maximum distance, which is always between the `firstCriticalPointIndex` and the `prevCriticalPointIndex` (the last one found). Initialized to -1.
 *
 * 2. Iteration:
 *    - Traverse the linked list using a `while` loop that continues as long as `currNode` and `currNode.next` are not null.
 *      This ensures that `currNode` always has both a `prevNode` and a `nextNode` to check for critical point conditions.
 *    - Inside the loop, determine if `currNode` is a local maxima (`currNode.val > prevNode.val && currNode.val > currNode.next.val`)
 *      or a local minima (`currNode.val < prevNode.val && currNode.val < currNode.next.val`).
 *
 * 3. Critical Point Handling:
 *    - If `currNode` is a critical point:
 *      - If `firstCriticalPointIndex` is -1, set `firstCriticalPointIndex` to the current `index`. This marks the first critical point.
 *      - Otherwise (if it's not the first critical point):
 *        - Update `minDistance` by taking the minimum of its current value and the distance between the current critical point and `prevCriticalPointIndex` (`index - prevCriticalPointIndex`).
 *        - Update `maxDistance` by calculating the distance between the current critical point and `firstCriticalPointIndex` (`index - firstCriticalPointIndex`).
 *      - In both cases, update `prevCriticalPointIndex` to the current `index`.
 *
 * 4. Pointers Update:
 *    - Advance the pointers: `prevNode = currNode`, `currNode = currNode.next`, and increment `index`.
 *
 * 5. Result Handling:
 *    - After the loop, if `firstCriticalPointIndex` is still -1 (no critical points found) or
 *      if `firstCriticalPointIndex` is equal to `prevCriticalPointIndex` (only one critical point found),
 *      return `[-1, -1]` as there are fewer than two critical points.
 *    - Otherwise, return `[minDistance, maxDistance]`.
 *
 * Time Complexity: O(N)
 * The algorithm makes a single pass through the linked list, performing constant-time operations for each node.
 * N is the number of nodes in the linked list.
 *
 * Space Complexity: O(1)
 * The algorithm uses a fixed number of variables regardless of the input list size.
 */

// Definition for singly-linked list.
// This constructor is typically provided by the LeetCode environment during execution.
// It's included here for completeness, making the file runnable standalone.
function ListNode(val, next) {
    this.val = (val === undefined ? 0 : val)
    this.next = (next === undefined ? null : next)
}

/**
 * @param {ListNode} head
 * @return {number[]}
 */
var nodesBetweenCriticalPoints = function(head) {
    // If the list has fewer than 3 nodes, it's impossible to have a critical point
    // because a critical point requires both a previous and a next node.
    if (!head || !head.next || !head.next.next) {
        return [-1, -1];
    }

    // Initialize pointers and index for traversal
    let prevNode = head;
    let currNode = head.next;
    let index = 1; // 0-based index: head is 0, head.next is 1, etc.

    // Variables to track critical points
    let firstCriticalPointIndex = -1; // Stores the index of the first critical point found
    let prevCriticalPointIndex = -1;  // Stores the index of the most recently found critical point

    // Variables to store min and max distances
    let minDistance = Infinity; // Initialize with a very large value to find the minimum
    let maxDistance = -1;       // Initialize with -1, will be updated if two CPs are found

    // Traverse the linked list
    // The loop condition `currNode && currNode.next` ensures that `currNode`
    // always has both a `prevNode` (which is `prevNode`) and a `nextNode` (`currNode.next`)
    // available for checking critical point conditions.
    while (currNode && currNode.next) {
        const nextNode = currNode.next;

        // Check if the current node is a local maxima or a local minima
        const isLocalMaxima = (currNode.val > prevNode.val && currNode.val > nextNode.val);
        const isLocalMinima = (currNode.val < prevNode.val && currNode.val < nextNode.val);

        if (isLocalMaxima || isLocalMinima) {
            // Found a critical point
            if (firstCriticalPointIndex === -1) {
                // This is the first critical point encountered
                firstCriticalPointIndex = index;
            } else {
                // This is a subsequent critical point.
                // Update minDistance: distance to the immediately preceding critical point
                minDistance = Math.min(minDistance, index - prevCriticalPointIndex);
                // Update maxDistance: distance from the first critical point to the current one.
                // This will always track the maximum possible distance because `firstCriticalPointIndex`
                // remains fixed and `index` is always increasing.
                maxDistance = index - firstCriticalPointIndex;
            }
            // Update the index of the most recently found critical point
            prevCriticalPointIndex = index;
        }

        // Move to the next nodes for the next iteration
        prevNode = currNode;
        currNode = nextNode;
        index++;
    }

    // After the traversal, check if at least two critical points were found.
    // If `firstCriticalPointIndex` is still -1, no critical points were found.
    // If `firstCriticalPointIndex` is equal to `prevCriticalPointIndex`, only one critical point was found.
    // In both these scenarios, it's impossible to calculate a distance between two *distinct* critical points.
    if (firstCriticalPointIndex === -1 || firstCriticalPointIndex === prevCriticalPointIndex) {
        return [-1, -1];
    }

    // If two or more critical points were found, return the calculated min and max distances.
    return [minDistance, maxDistance];
};

File: solutions/2026-08-31/python/2026-08-31-find-the-minimum-and-maximum-number-of-nodes-between-critical-points.py
# Problem: Find the Minimum and Maximum Number of Nodes Between Critical Points
# Link: https://leetcode.com/problems/find-the-minimum-and-maximum-number-of-nodes-between-critical-points/
#
# Problem Summary:
# Given a singly linked list, find "critical points" which are local maxima
# or local minima. A node is a local maxima if its value is strictly greater
# than its previous and next nodes. A node is a local minima if its value is
# strictly smaller than its previous and next nodes. The first and last nodes
# cannot be critical points. Return an array [minDistance, maxDistance]
# representing the minimum and maximum distances between any two distinct
# critical points. If fewer than two critical points exist, return [-1, -1].
#
# Approach Explanation:
# 1. Initialize an empty list `critical_points_indices` to store the 0-based
#    indices of all critical points found during the traversal.
# 2. Use three pointers (`prev_node`, `curr_node`, `next_node`) and an `index`
#    counter (starting from 1, as `head` is at index 0) to iterate through the
#    linked list. `prev_node` starts at `head`, `curr_node` starts at `head.next`.
# 3. The traversal loop continues as long as `curr_node` and `curr_node.next` are
#    not None. This ensures that for every `curr_node` being examined, there is
#    both a `prev_node` and a `next_node` available, which is a prerequisite
#    for a node to be a critical point.
# 4. Inside the loop, check if the `curr_node` meets the criteria for a local
#    maxima (`curr_node.val > prev_node.val` AND `curr_node.val > next_node.val`)
#    or a local minima (`curr_node.val < prev_node.val` AND `curr_node.val < next_node.val`).
# 5. If `curr_node` is a critical point, add its current `index` to the
#    `critical_points_indices` list.
# 6. After checking, advance the pointers: `prev_node` becomes `curr_node`,
#    `curr_node` becomes `next_node`, and `index` is incremented.
# 7. Once the traversal is complete, check the length of `critical_points_indices`.
#    If it contains fewer than two elements, it means not enough critical points
#    were found to calculate distances, so return `[-1, -1]`.
# 8. Otherwise, calculate `maxDistance` as the difference between the index of the
#    last critical point and the first critical point in the list.
# 9. Calculate `minDistance` by iterating through `critical_points_indices` and
#    finding the minimum difference between any two adjacent critical points.
#    Initialize `minDistance` to a very large value (e.g., `float('inf')`) and
#    update it with `min(minDistance, critical_points_indices[i+1] - critical_points_indices[i])`.
# 10. Finally, return the calculated `[minDistance, maxDistance]`.
#
# Time Complexity:
# O(N), where N is the number of nodes in the linked list.
# We perform a single pass through the linked list to identify all critical points.
# This operation takes O(N) time. Subsequently, we iterate through the list of
# critical point indices (which can have at most N/2 elements) to calculate the
# minimum distance. This takes O(C) time, where C is the number of critical points,
# which is also O(N) in the worst case. Therefore, the overall time complexity is O(N).
#
# Space Complexity:
# O(N), where N is the number of nodes in the linked list.
# In the worst case, every other node could be a critical point (e.g., a list with
# alternating local maxima and minima). This would require storing approximately
# N/2 indices in the `critical_points_indices` list. Thus, the space complexity is O(N).

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def nodesBetweenCriticalPoints(self, head: ListNode) -> list[int]:
        # List to store the 0-based indices of all critical points found.
        critical_points_indices = []
        
        # We need three pointers: prev_node, curr_node, and curr_node.next
        # to determine if curr_node is a critical point.
        # The first and last nodes cannot be critical points, as they lack
        # a previous or next node respectively.
        
        # Initialize prev_node to the head of the list.
        prev_node = head
        # Initialize curr_node to the second node.
        # If head is None or head.next is None, curr_node will be None,
        # and the loop condition will correctly handle this (no critical points possible).
        curr_node = head.next
        # Initialize index. We use 0-based indexing for convenience.
        # head is at index 0, curr_node (head.next) is at index 1.
        index = 1
        
        # Iterate through the list. The loop condition `curr_node and curr_node.next`
        # ensures that `curr_node` is not the last node and can have a `next_node`.
        # This way, `prev_node`, `curr_node`, and `curr_node.next` are always valid
        # for checking critical point conditions.
        while curr_node and curr_node.next:
            # Get the next node in the sequence.
            next_node = curr_node.next
            
            # Check for local maxima: curr_node.val is strictly greater than its neighbors.
            is_local_maxima = curr_node.val > prev_node.val and curr_node.val > next_node.val
            # Check for local minima: curr_node.val is strictly smaller than its neighbors.
            is_local_minima = curr_node.val < prev_node.val and curr_node.val < next_node.val
            
            # If curr_node is either a local maxima or a local minima, it's a critical point.
            if is_local_maxima or is_local_minima:
                critical_points_indices.append(index)
            
            # Move pointers to the next nodes for the next iteration.
            prev_node = curr_node
            curr_node = next_node
            index += 1
            
        # After traversing the list, check the number of critical points found.
        # If there are fewer than two critical points, return [-1, -1].
        if len(critical_points_indices) < 2:
            return [-1, -1]
        
        # Calculate maxDistance: the distance between the first and last critical points.
        # Since critical_points_indices is sorted, this is simply the difference
        # between the last and first element.
        max_distance = critical_points_indices[-1] - critical_points_indices[0]
        
        # Calculate minDistance: the minimum distance between any two adjacent critical points.
        # Initialize min_distance to a very large value.
        min_distance = float('inf')
        # Iterate through the critical_points_indices list to find the minimum difference
        # between consecutive critical points.
        for i in range(len(critical_points_indices) - 1):
            min_distance = min(min_distance, critical_points_indices[i+1] - critical_points_indices[i])
            
        # Return the calculated minDistance and maxDistance.
        return [min_distance, max_distance]